### PR TITLE
Edit: Track persistence of edits and insertions

### PR DESCRIPTION
**Pull request by umpox**
## Description

This PR makes `PersistenceTracker` generic and uses it for `Edit`

In its current state, this means we get:
- Accurate tracking for _insertion_ edits (Document, Test, Generate)
- Partially accurate tracking for _edit_ edits (Fix, Instruction-based Edit)

The latter is only partially accurate as it tracks the entire selectionRange, but an Edit might only update part of this range.

Will address this in a follow up as part of https://github.com/sourcegraph/cody/issues/3513

## Test plan

**Insertion edit**
1. Make an insertion edit
2. Delete nothing, check that telemetry is sent with correct `difference` after 30 seconds
3. Repeat but by deleting part of the insertion and then the full insertion, validate the `difference` each time.

**Editing edit**
1. Make an instruction-based edit
2. Delete nothing, check that telemetry is sent with correct `difference` after 30 seconds
3. Repeat but by deleting part of the insertion and then the full insertion, validate the `difference` each time.
4. Note that this `difference` may not accurately reflect Cody's changes

### CODE REVIEW

File: vscode/src/common/persistence-tracker/index.ts [hunk 1]
```diff
@@ -26,19 +26,25 @@ interface TrackedCompletion {
...
     constructor(
         workspace: Pick<
             typeof vscode.workspace,
             'onDidChangeTextDocument' | 'onDidRenameFiles' | 'onDidDeleteFiles'
-        > = vscode.workspace
+        >,
+        public logger: {
+            onRemoved: (event: PersistenceRemovedEventPayload<T>) => void
+            onPresent: (event: PersistencePresentEventPayload<T>) => void
+        }
```
Comment (umpox) [inline vscode/src/common/persistence-tracker/index.ts:47]: Change in this file: Consumer now provides the `logger` and handles the telemetry events itself

File: vscode/src/common/persistence-tracker/index.ts [hunk 2]
```diff
@@ -51,116 +57,105 @@ export class PersistenceTracker implements vscode.Disposable {
...
-        const textLines = lines(insertText)
-        const latestRange = new vscode.Range(
-            insertRange.start.line,
-            insertRange.start.character,
-            insertRange.end.line + textLines.length - 1,
-
-            textLines.length > 1
-                ? textLines.at(-1)!.length
-                : insertRange.end.character + textLines[0].length
-        )
```
Comment (umpox) [inline vscode/src/common/persistence-tracker/index.ts:76]: Change in this file: Let the consumer do their own calculation and here we can just assume they provide the correct `insertRange`.

File: vscode/src/common/persistence-tracker/index.ts [hunk 3]
```diff
@@ -26,19 +26,25 @@ interface TrackedCompletion {
     insertText: string
     insertRange: vscode.Range
     latestRange: vscode.Range
+    metadata?: PersistenceEventMetadata
```
Comment (umpox) [inline vscode/src/common/persistence-tracker/index.ts:29]: Change in this file: We support attaching arbitrary metadata that will be sent alongside `onRemoved` and `onPresent`. Used in Edit for tracking persistence across different edit commands (e.g. test, document, etc)

**Review by philipp-spiess — APPROVED** [review]
Chef Thumb

**Pull request merged**